code/datums/components/vis_block.dm

```diff
@@ -0,0 +1,240 @@
+	if(!isturf(atom_parent.loc) || !opacity)
+		if(actively_blocking)
+			actively_blocking = FALSE
+			us_finder.moveToNullspace()
+		return
+	if(!actively_blocking)
+		actively_blocking = TRUE
+		us_finder.forceMove(atom_parent.loc)
+	var/turf/home = atom_parent.loc
+
+	/// This is the darkness everything else will be masking out
+	var/mutable_appearance/darkness_base = mutable_appearance('icons/obj/doors/airlocks/tall/airlock_darkness.dmi', "[wall_junction]", offset_spokesman = atom_parent, plane = DARKNESS_MASK_PLANE)
+	darkness_base.pixel_w = -5
+	partial_darkness = FALSE
+	for(var/check_dir in GLOB.cardinals)
+		if(!(check_dir & (dir | REVERSE_DIR(dir))))
+			// If we are not up against one of our own own, OR a wall
+			if(check_dir & (wall_junction|us_junction))
+				continue
+			// AND we have holes in our darkness (an "inside")
+			if(inner_transparent_dirs == NONE)
+				continue
+			// Then we draw a mask on ourselves, on our OWN turf, to allow say the turf below some vertical headroom
+			darkness_base.add_overlay(mutable_appearance('icons/obj/doors/airlock_mask.dmi', "[edge_dir_mask]_dark_edge_[dir2text(check_dir)]", HIGHEST_GAME_LAYER, offset_spokesman = atom_parent, plane = DARKNESS_MASK_PLANE))
+			continue
+		var/turf/in_direction = get_step(home, check_dir)
+		if(!in_direction)
+			continue
+
+		// Base mask applied to darkness, masks out only the side we can see (since we overlay it on the turf facing said side)
+		var/mutable_appearance/mask = mutable_appearance('icons/obj/doors/airlock_mask.dmi', dir_mask, HIGHEST_GAME_LAYER, offset_spokesman = atom_parent, plane = DARKNESS_MASK_PLANE)
+		mask = make_ma_directional(mask)
+		mask.dir = check_dir
+		mask.pixel_w = -5
+		switch(check_dir)
+			if(NORTH)
+				mask.pixel_y = -32
+			if(SOUTH)
+				mask.pixel_y = 32
+			if(EAST)
+				mask.pixel_x = -32
+			if(WEST)
+				mask.pixel_x = 32
+
+		// These handle connecting edges with walls
+		// (since they have different bounds and I want to show walls say, going in slightly in some cases without doing the same to connected blockers)
+		if(wall_junction & NORTH)
+			mask.add_overlay(mutable_appearance(mask.icon, "[edge_dir_mask]_edge_north"))
+		if(wall_junction & SOUTH)
+			mask.add_overlay(mutable_appearance(mask.icon, "[edge_dir_mask]_edge_south"))
+		if(wall_junction & EAST)
+			mask.add_overlay(mutable_appearance(mask.icon, "[edge_dir_mask]_edge_east"))
+		if(wall_junction & WEST)
+			mask.add_overlay(mutable_appearance(mask.icon, "[edge_dir_mask]_edge_west"))
+		// If our inside is transparent in this case, draw the standard darkness
+		// Using a half alpha and BLEND_SUBTRACT
+		// This way if both stack we'll get darkness and opacity, and if they don't nothing will happen
+		// Also only apply once per turf to avoid breaking on overlap
+		if(check_dir & inner_transparent_dirs)
+			partial_darkness = TRUE
+			if(!HAS_TRAIT(home, TRAIT_INNER_DARKNESS) || HAS_TRAIT_FROM(home, TRAIT_INNER_DARKNESS, WEAKREF(src)))
```

Comment:
If the component itself is the only source of the trait, this check is likely redundant as it'll always pass. Are there other sources of this trait around or any other reason this has to be a thing?

Reply:
TRAIT_INNER_DARKNESS could be applied by other instances of this component.
But also, this HAS_TRAIT_FROM is bad because sources should be string values, not datums iirc. This needs to be changed to REF(src)

Reply:
This isn't a dupable component iirc, so there should be only one instance at a time, but I kind of get it.

Reply:
more then one component can be on the turf at a time. Honestly this could just be a var off turf and the component but I was nightcoding, accepting input on this topic

Reply:
Oh, I was under the wrong impression that the component was attahed to the turf while reading this line. All good then.